src/sif/autocost.cc

```diff
@@ -590,19 +653,40 @@ Cost AutoCost::TransitionCostReverse(const uint32_t idx,
         turn_cost += 0.5f;
     }
 
+    float seconds = turn_cost;
+    bool is_turn = false;
+    bool has_left = (edge->turntype(idx) == baldr::Turn::Type::kLeft ||
+                     edge->turntype(idx) == baldr::Turn::Type::kSharpLeft);
+    bool has_right = (edge->turntype(idx) == baldr::Turn::Type::kRight ||
+                      edge->turntype(idx) == baldr::Turn::Type::kSharpRight);
+    bool has_reverse = edge->turntype(idx) == baldr::Turn::Type::kReverse;
+
     // Separate time and penalty when traffic is present. With traffic, edge speeds account for
     // much of the intersection transition time (TODO - evaluate different elapsed time settings).
     // Still want to add a penalty so routes avoid high cost intersections.
-    float seconds = turn_cost;
-    bool is_turn = false;
-    if (edge->turntype(idx) == baldr::Turn::Type::kLeft ||
-        edge->turntype(idx) == baldr::Turn::Type::kSharpLeft ||
-        edge->turntype(idx) == baldr::Turn::Type::kRight ||
-        edge->turntype(idx) == baldr::Turn::Type::kSharpRight) {
+    if (has_left || has_right || has_reverse) {
       seconds *= edge->stopimpact(idx);
       is_turn = true;
     }
 
+    if (node->drive_on_right()) {
```

Comment:
instead of duplicating this a bunch, could we move it into `base_transition_cost` and then have different costings opt in and out of it by setting a boolean member variable that controls whether it happens or not. then costings like pedestrian can set the flag `DynamicCost::penalize_uturns = false` and so when they call `base_transition_cost` it doesnt get applied but when AutoCost calls it it will have the member variable set to true and it will be applied. i'm just trying to lessen the burdon of maintaining this complexity by sequestering it to one place in the code base :smile:

Reply:
@kevinkreiser hmmm may make `TransitionCostReverse` confusing cause it does not have the same arguments at TransitionCost.  We still have not resolved https://github.com/valhalla/valhalla/issues/2896 yet.  

```
Cost AutoCost::TransitionCostReverse(const uint32_t idx,
                                     const baldr::NodeInfo* node,
                                     const baldr::DirectedEdge* pred,
                                     const baldr::DirectedEdge* edge,
                                     const bool has_measured_speed,
                                     const InternalTurn internal_turn) const {
```

Reply:
If we ever get that resolved then maybe we could move it there.

Reply:
ok cool well then instead of base_transition_cost could we at least turn it into a method on DynamicCost for forward and reverse and have everything share it so at least its all in one place?

Reply:
@kevinkreiser ok updated.

Reply:
i just wish there were a way we could take this block out of TransitionCostReverse and TransitionCost and put them into DynamicCost and use them from there. id really like it if we could whittle down the costings into very slight tweaks on DynamicCost base class functionality so there weren't so much copy pasta and code to maintain, so that when we add a costing feature it gets added to all places more easily etc.

could we pull this and the forward direction out into methods like:

```c++
//pass seconds by reference so it can be modified and inline the functions
AddUturnPenalty(node, edge, has_reverse, has_left, has_right, internal_turn seconds);
AddUturnPenaltyReverse(node, edge, has_reverse, has_left, has_right, internal_turn, seconds);
```

Reply:
@kevinkreiser sure will do

Reply:
@kevinkreiser was able to do this with one function.  if we need to split it out later we can do so.

``
  /**
   * Adds a penalty to 3 types of uturns.  1) uturn on a short, internal edge 2) uturn at a node
   * 3) pencil point uturn. Note that motor_scooter and motorcycle costing models do not penalize
   * uturns on a short, internal edge; hence, the boolean penalize_internal_uturns.
   * @param  idx          Directed edge local index
   * @param  node         Node (intersection) where transition occurs.
   * @param  edge         Directed edge (the to edge)
   * @param  has_reverse  Did we perform a reverse?
   * @param  has_left     Did we make a left (left or sharp left)
   * @param  has_right    Did we make a right (right or sharp right)
   * @param  penalize_internal_uturns   Do we want to penalize uturns on a short, internal edge
   * @param  internal_turn              Did we make an turn on a short internal edge.
   * @param  seconds      Time.
   */
  inline void AddUturnPenalty(const uint32_t idx,
                              const baldr::NodeInfo* node,
                              const baldr::DirectedEdge* edge,
                              const bool has_reverse,
                              const bool has_left,
                              const bool has_right,
                              const bool penalize_internal_uturns,
                              const InternalTurn internal_turn,
                              float& seconds) const
```